### PR TITLE
feat(axes): option to fit domain to list of annotation `SpecIds`

### LINKED ISSUE #1411: Axis domain fit option to include also annotation's data
**Is your feature request related to a problem? Please describe.**
While working on the new threshold feature in Lens, using the annotation components from this library, we found the need to use an automatic fit that would include also annotations in the domain calculation.

**Describe the solution you'd like**
Having a `fit` option to fit the data in the chart, including the annotation one.

**Describe alternatives you've considered**
Doing it manually on consumer side.

**Kibana Cross Issues**
https://github.com/elastic/kibana/issues/112795
https://github.com/elastic/kibana/issues/129240

### PR DESCRIPTION
**Pull request by nickofthyme**
## Summary

Adds the `AxisSpec.domain.includeDataFromIds` option to `YDomainBase`. This optional allows specifying a series of `SpecIds` to include into the domain calculation, respective of their `groupId`. Currently, only `LineAnnotation` and `RectAnnotation` specs are supported, everything else is already included in the domain automatically. Setting `domain.max` or `domain.min` will override this functionality.

Apr-06-2022 11-09-54

See pr demo story [here](https://ech-e2e-ci--pr-1641-c6onxe2b.web.app/?path=/story/axes--fit-domain&globals=theme:light&knob-Domain%20padding%20unit=domainRatio&knob-constrain%20padding=true&knob-dataset=positive&knob-domain%20padding=0.1&knob-fit%20Y%20domain%20to%20annotations=true&knob-fit%20Y%20domain%20to%20data=true&knob-thesholds=900,%201100,1300&knob-Specs%20to%20fit%20(yDomain)[0]=theshold&knob-Specs%20to%20fit%20(yDomain)[1]=rect&knob-thesholds%20-%20line[0]=1400&knob-theshold%20-%20rect={%22y0%22:null,%22y1%22:900})

## Issues

fix #1411

### Checklist

- [x] The proper **chart type** label has been added (e.g. `:xy`, `:partition`)
- [x] The proper **feature** labels have been added (e.g. `:interactions`, `:axis`)
- [x] All related issues have been linked (i.e. `closes #123`, `fixes #123`)
- [x] New public API exports have been added to `packages/charts/src/index.ts`
- [x] Unit tests have been added or updated to match the most common scenarios
- [x] The proper documentation and/or storybook story has been added or updated

### CODE REVIEW

**Comment by nickofthyme** [comment]
@markov00 I added the rect annotation finite values to the domain. The one question I have is when the domain `min` is controlled by `y0` or the `min` by `y1`, the rect annotation covers the full area of the chart with no real indication that it's an annotation without the some type of padding. See [example](https://ech-e2e-ci--pr-1641-c6onxe2b.web.app/?path=/story/axes--fit-domain&globals=theme:light&knob-dataset=both&knob-fit%20Y%20domain%20to%20data=true&knob-fit%20Y%20domain%20to%20annotations=true&knob-constrain%20padding=true&knob-domain%20padding=0&knob-Domain%20padding%20unit=domainRatio&knob-thesholds%20-%20line[0]=200&knob-theshold%20-%20rect={%22y0%22:-200,%22y1%22:null}).

Screen Recording 2022-04-07 at 05 38 48 PM

I wonder if we should filter the `y0` rect annotation values that are below the computed min and `y1` values above the computed max.

**Comment by markov00** [comment]
> @markov00 I added the rect annotation finite values to the domain. The one question I have is when the domain `min` is controlled by `y0` or the `max` by `y1`, the rect annotation covers the full area of the chart with no real indication that it's an annotation without the some type of padding. See [example](https://ech-e2e-ci--pr-1641-c6onxe2b.web.app/?path=/story/axes--fit-domain&globals=theme:light&knob-dataset=both&knob-fit%20Y%20domain%20to%20data=true&knob-fit%20Y%20domain%20to%20annotations=true&knob-constrain%20padding=true&knob-domain%20padding=0&knob-Domain%20padding%20unit=domainRatio&knob-thesholds%20-%20line%5B0%5D=200&knob-theshold%20-%20rect=%7B%22y0%22:-200,%22y1%22:null%7D).
>  
> I wonder if we should filter the `y0` rect annotation values that are below the computed min and `y1` values above the computed max.

I don't see this as a problem. If I'm adding an annotation with a data domain that is bigger then the current chart data domain, and I configured my chart to include the annotation domain into the data once, I don't see this as an issue: a rectangle area that covers entirely the chart works fine for me, it doesn't need anything special

File: storybook/preview.ts [hunk 1]
```diff
@@ -7,15 +7,9 @@
...
 import { storybookParameters } from './parameters';
-import { preloadIcons } from './preload_icons';
 import { StoryWrapper } from './story_wrapper';
 
 import './style.scss';
 
-if (process.env.VRT) {
-  preloadIcons();
-  document.querySelector('html')?.classList.add('disable-animations');
-}
```
Comment (markov00) [inline storybook/preview.ts:18]: Great, I was doing the same

Comment (nickofthyme) [reply storybook/preview.ts:18]: Yeah I did this in the playwright work and was irritated it was still there 😄

File: packages/charts/src/chart_types/xy_chart/domains/y_domain.ts [hunk 2]
```diff
@@ -57,8 +68,10 @@ function mergeYDomainForGroup(
   if (isStacked && stackMode === StackMode.Percentage) {
     mergedDomain = computeContinuousDataDomain([0, 1], type, customDomain);
   } else {
-    const stackedDomain = computeYDomain(stacked, hasZeroBaselineSpecs, type, newCustomDomain);
-    const nonStackedDomain = computeYDomain(nonStacked, hasZeroBaselineSpecs, type, newCustomDomain);
+    const annotationData =
+      newCustomDomain.fitAnnotations ?? newCustomDomain.fit ? annotationYValueMap.get(groupId) ?? [] : [];
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply packages/charts/src/chart_types/xy_chart/domains/y_domain.ts]: See 04e1ed6 and 02ba736

File: packages/charts/src/chart_types/xy_chart/utils/specs.ts [hunk 3]
```diff
@@ -325,6 +325,13 @@ export interface YDomainBase {
    * @defaultValue false
    */
   fit?: boolean;
+  /**
+   * Whether to fit the domain to annotstion data values.
+   *
+   * Setting `max` or `min` will override this functionality.
+   * @defaultValue YDomainBase.fit
+   */
+  fitAnnotations?: boolean;
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply packages/charts/src/chart_types/xy_chart/utils/specs.ts]: Fixed errors from this in 02ba736

**Review by markov00 — APPROVED** [review]
Tested locally,  everything works like perfectly, I have also tested with the `yNice` prop enabled and it follows that correctly.

Please update the PR description with the changes you have done on the naming

**Pull request merged**